### PR TITLE
libjose

### PR DESCRIPTION
**Pull request by l1h3r**
# Description of change

libjose is a JSON Web Token library built using crypto.rs.

At the moment this is needed for DIDComm but eventually will be used for SIOP DID and [Verification Methods](https://github.com/iotaledger/identity.rs/issues/118).

Start Here:

- [rfcs](https://github.com/iotaledger/identity.rs/blob/feat/libjose/libjose/src/lib.rs#L12) (overview. dont need to read all of these)
- [examples](https://github.com/iotaledger/identity.rs/tree/feat/libjose/libjose/examples)
- [tests](https://github.com/iotaledger/identity.rs/tree/feat/libjose/libjose/tests)

## Type of change

- [x] Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

[tests](https://github.com/iotaledger/identity.rs/tree/feat/libjose/libjose/tests)

## Change checklist

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: libjose/src/utils/serde.rs [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+use core::str;
+
+use crate::error::Error;
+use crate::error::Result;
+use crate::jose::JoseHeader;
+use crate::jwe::JweHeader;
+use crate::jws::JwsHeader;
+use crate::lib::*;
```
[the remark this answers is not in the document]
Comment (JelleMillenaar) [reply libjose/src/utils/serde.rs:11]: We would like to move IOTA Identity towards NO_STD for Rust embedded in the future. We try to minimize standard library usage.

File: libjose/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+rand = { version = "0.7", default-features = false, features = ["getrandom"] }
+rsa = { version = "0.3", default-features = false }
+sha-1 = { version = "0.9", default-features = false }
+sha2 = { version = "0.9", default-features = false }
+subtle = { version = "2.4", default-features = false }
+
+[dependencies.crypto]
+git = "https://github.com/iotaledger/crypto.rs"
+branch = "libjose"
+# path = "../../crypto.rs"
```
[the remark this answers is not in the document]
Comment (l1h3r) [reply libjose/Cargo.toml:28]: committed by mistake. will remove 👍

File: libjose/rust-toolchain [hunk 3]
```diff
@@ -0,0 +1 @@
+stable
```
[the remark this answers is not in the document]
Comment (l1h3r) [reply libjose/rust-toolchain]: originally I was building this library with nightly features and this sets the default toolchain.

> Will projects that use libjose require stable?

Yeah, not locked to nightly anymore and currently supporting `no_std` with the `alloc` crate

File: libjose/src/jwe/recipient.rs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+impl<'a, T> From<T> for Recipient<'a>
+where
+  T: Into<Secret<'a>>,
+{
+  fn from(other: T) -> Self {
+    Self::new(other)
+  }
+}
+
+impl<'a, T> From<(T, &'a JweHeader)> for Recipient<'a>
```
[the remark this answers is not in the document]
Comment (l1h3r) [reply libjose/src/jwe/recipient.rs:48]: Used a bit in the tests but yeah, not needed 👍

File: libjose/src/jwk/key.rs [hunk 5]
```diff
@@ -0,0 +1,558 @@
...
+          [..] => return Err(Error::KeyError("multi prime keys are not supported")),
+        };
+
+        Ok(Jwk::from_params(JwkParamsRsa {
+          n: encode_b64(secret.n().to_bytes_be()),
+          e: encode_b64(secret.e().to_bytes_be()),
+          d: Some(encode_b64(secret.d().to_bytes_be())),
+          p: Some(encode_b64(p.to_bytes_be())),
+          q: Some(encode_b64(q.to_bytes_be())),
+          dp: Some(encode_b64(&[])), // TODO: FIXME
```
[the remark this answers is not in the document]
Comment (l1h3r) [reply libjose/src/jwk/key.rs:528]: Yes

File: libjose/src/utils/crypto/random.rs [hunk 6]
```diff
@@ -0,0 +1,13 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::error::Result;
+use crate::lib::*;
```
[the remark this answers is not in the document]
Comment (l1h3r) [reply libjose/src/utils/crypto/random.rs:5]: This is needed for `Vec` and `vec!` when not using `std` (`--no-default-features --features alloc`).

Probably don't need a wildcard import though

File: libjose/src/jwm/attributes.rs [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+    self.reply_to.as_deref()
+  }
+
+  /// Sets a value for the reply to atribute (reply_to).
+  pub fn set_reply_to(&mut self, value: impl IntoIterator<Item = impl Into<String>>) {
+    self.reply_to = Some(value.into_iter().map(Into::into).collect());
+  }
+}
+
+impl<T> Deref for JwmAttributes<T> {
```
[the remark this answers is not in the document]
Comment (l1h3r) [reply libjose/src/jwm/attributes.rs:191]: No but you **do** need access to `claims`. Will replace this with explicit methods 👍

File: libjose/src/utils/crypto/key_repr.rs [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+use crate::error::Result;
+use crate::jwk::EcCurve;
+use crate::jwk::EcxCurve;
+use crate::jwk::EdCurve;
+use crate::jwk::Jwk;
+use crate::jwk::JwkParamsEc;
+use crate::jwk::JwkParamsOct;
+use crate::jwk::JwkParamsOkp;
+use crate::jwk::JwkParamsRsa;
+use crate::lib::*;
```
[the remark this answers is not in the document]
Comment (l1h3r) [reply libjose/src/utils/crypto/key_repr.rs:16]: needed for `no_std`

File: libjose/src/jwt/header_set.rs [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+use crate::error::Result;
+use crate::jwe::JweAlgorithm;
+use crate::jwe::JweCompression;
+use crate::jwe::JweEncryption;
+use crate::jwe::JweHeader;
+use crate::jwk::Jwk;
+use crate::jws::JwsAlgorithm;
+use crate::jws::JwsHeader;
+use crate::jwt::JwtHeader;
+use crate::lib::*;
```
[the remark this answers is not in the document]
Comment (l1h3r) [reply libjose/src/jwt/header_set.rs:17]: needed for `no_std`

**Review by JelleMillenaar — APPROVED** [review]
Happy to merge in its current state if you are comfortable with it since it will be moved to crypto.rs later anyway.

**Pull request merged**